### PR TITLE
Don't format regions inside inactive conditionals

### LINKED ISSUE #62612: False IDE0055 (Fix formatting) reported when using multiple TargetFrameworks
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/False-IDE0055-Fix-formatting-reported-/10093052)._

---
[severity:It bothers me. A fix would be nice]
I am compiling a class for multiple TargetFrameworks. Some parts of the class are only compiled for certain TargetFrameworks. Correct formatting of the documents is enforced via .editorconfig and `EnforceCodeStyleInBuild`.

It seems, the desired formatting of `#region`-statements depends on whether the code is active or not. So no matter how I try to fix the formatting - the build will always raise a warning for one of the frameworks.

In the screenshot it is nicely visible, that line 6 has the same formatting as line 17, yet the formatting of line 6 is flagged.

Workaround: Avoid usage of `#region`-statements in code that is compiled conditionally.

---
#### Feedback Bot on 7/12/2022, 07:46 PM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #62612

### CODE REVIEW

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs [hunk 1]
```diff
@@ -99,7 +99,20 @@ protected override LineColumnRule GetLineColumnRuleBetween(SyntaxTrivia trivia1,
...
+                    }
+
+                    if (!previous.IsKind(SyntaxKind.DisabledTextTrivia))
+                    {
+                        return LineColumnRule.PreserveLinesWithDefaultIndentation(lines);
+                    }
+                    else
+                    {
+                        return LineColumnRule.Preserve;
+                    }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:115]: this likely warrants docs :)

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs [hunk 2]
```diff
@@ -99,7 +99,20 @@ protected override LineColumnRule GetLineColumnRuleBetween(SyntaxTrivia trivia1,
...
                 if (trivia2.IsKind(SyntaxKind.RegionDirectiveTrivia) || trivia2.IsKind(SyntaxKind.EndRegionDirectiveTrivia))
                 {
-                    return LineColumnRule.PreserveLinesWithDefaultIndentation(lines);
+                    var previous = trivia2;
+                    while ((previous = previous.GetPreviousTrivia(previous.SyntaxTree, CancellationToken.None)).IsKind(SyntaxKind.WhitespaceTrivia))
+                    {
+                        // skip whitespace trivia.
+                    }
+
+                    if (!previous.IsKind(SyntaxKind.DisabledTextTrivia))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:108]: nit, consider inverting.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs [hunk 3]
```diff
@@ -99,7 +99,20 @@ protected override LineColumnRule GetLineColumnRuleBetween(SyntaxTrivia trivia1,
 
                 if (trivia2.IsKind(SyntaxKind.RegionDirectiveTrivia) || trivia2.IsKind(SyntaxKind.EndRegionDirectiveTrivia))
                 {
-                    return LineColumnRule.PreserveLinesWithDefaultIndentation(lines);
+                    var previous = trivia2;
+                    while ((previous = previous.GetPreviousTrivia(previous.SyntaxTree, CancellationToken.None)).IsKind(SyntaxKind.WhitespaceTrivia))
+                    {
+                        // skip whitespace trivia.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:105]: needs more explanation :)

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:105]: what about end-of-line trivia?

Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:105]: > what about end-of-line trivia?

The following test just passes:

```csharp
        [Fact]
        public async Task TestRegion2()
        {
            var testCode = @"
class MyClass
{
#if true
    public void M()
    {
[||]#region ABC
        System.Console.WriteLine();
[||]#endregion
    }
#else
    public void M()
    {
#region ABC
        System.Console.WriteLine();
#endregion
    }
#endif
}
";
            var fixedCode = @"
class MyClass
{
#if true
    public void M()
    {
        #region ABC
        System.Console.WriteLine();
        #endregion
    }
#else
    public void M()
    {
#region ABC
        System.Console.WriteLine();
#endregion
    }
#endif
}
";
            await Verify.VerifyCodeFixAsync(testCode, fixedCode);
        }
```

(I'll definitely commit and push it)

`'#region'.GetPreviousTrivia(..)` is not an end of line trivia for this case though.

Do you have an interesting test case where the tree will have end of line trivia?

I found a niche case where my logic is incorrect apart from EOL trivia:

```csharp
#region OUTER
        #region ABC
        System.Console.WriteLine();
        #endregion
#endregion
```

Here, after we go back consuming whitespace trivia, we don't find a disabled text, but we are in disabled text.

Is there a compiler API for the "Is this position conditionally excluded?" question?

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:105]: does IsInInactiveRegion work?

Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:105]: @CyrusNajmabadi It doesn't 😕 

https://github.com/dotnet/roslyn/blob/75bbc51a05b71d8a07c861201660f55974909845/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/SyntaxTreeExtensions.cs#L477-L538

That seems a hard-to-answer question on the IDE side.

Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:105]: > That seems a hard-to-answer question on the IDE side.

Not really. I think I see how I can do it.

Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs]: @CyrusNajmabadi This is ready for another look now :)

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by CyrusNajmabadi** [comment]
Currently failing test: Microsoft.CodeAnalysis.CSharp.UnitTests.Formatting.FormattingEngineTriviaTests.MixAll

**Comment by Youssef1313** [comment]
> Currently failing test: Microsoft.CodeAnalysis.CSharp.UnitTests.Formatting.FormattingEngineTriviaTests.MixAll

Fixed. Not an ideal fix though, just falling back to the original behavior for edge cases that are harder to handle.

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Can you take a look please? Thanks!

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs [hunk 4]
```diff
@@ -99,6 +99,38 @@ protected override LineColumnRule GetLineColumnRuleBetween(SyntaxTrivia trivia1,
...
+                    {
+                        if (previous.IsKind(SyntaxKind.DisabledTextTrivia))
+                        {
+                            return LineColumnRule.Preserve;
+                        }
+                        else if (previous.IsKind(SyntaxKind.EndIfDirectiveTrivia))
+                        {
+                            // We cannot go back further to determine if we are in a disabled region or not.
+                            // We assume we are in active region, but that might not be correct.
+                            break;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:119]: i'm curious why you can't go further back?  can't this go to the corresponding `#if` and continue searching?

**Review by CyrusNajmabadi — APPROVED** [review]
tentative approval.  but would like answer to question posed.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs (hunk 4 above)
Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:119]: Ah, it's possible, something like this (pseudo-code):

```csharp
if (found #endif)
{
    int endifCount = 1;
    int ifCount = 0;
    while (go back)
    {
        if (found #endif) endifCount++;
        else if (found #if) ifCount++;
        if (ifCount == endifCount) break;
    }
}
```

It's a niche scenario though, and we might have to go back too much.

Comment (Youssef1313) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Engine/Trivia/CSharpTriviaFormatter.cs:119]: If it's important to handle this, I'll go ahead and do it.

**Pull request merged**